### PR TITLE
Add support for per e-info rwlock and add more debug

### PR DESCRIPTION
**Pull request by megachips**

### CODE REVIEW

File: emlog.c [hunk 1]
```diff
@@ -256,10 +258,16 @@ static char * read_from_emlog(struct emlog_info * einfo, size_t * length,
...
+
     /* is the user trying to read data that has already scrolled off? */
     if (*offset < einfo->offset)
         *offset = einfo->offset;
-
+    if (emlog_debug)
+        pr_debug("New Offset: %lld\n", *offset);
     /* is the user trying to read past EOF? */
     if (*offset >= EMLOG_FIRST_EMPTY_BYTE(einfo))
         return NULL;
```
Comment (ahippo) [inline emlog.c]: Missing `read_unlock()`?

File: emlog.c [hunk 2]
```diff
@@ -268,11 +276,15 @@ static char * read_from_emlog(struct emlog_info * einfo, size_t * length,
...
     start_point = einfo->read_point + (*offset - einfo->offset);
+    if (emlog_debug)
+        pr_debug("Start point: %d\n", start_point);
     start_point = start_point % einfo->size;
-
+    if (emlog_debug)
+        pr_debug("Start point: %d\n", start_point);
     /* allocate memory to return */
     if ((retval = kmalloc(sizeof(char) * remaining, GFP_KERNEL)) == NULL)
         return NULL;
```
Comment (ahippo) [inline emlog.c]: Missing `read_unlock()`?

File: emlog.h [hunk 3]
```diff
@@ -39,6 +39,7 @@ struct emlog_info {
     size_t write_point;         /* Offset in circ. buffer of newest data */
     loff_t offset;              /* Byte number of read_point in the stream */
     struct emlog_info *next;
+    rwlock_t rwlock;            /* per e-info rwlock */
```
Comment (ahippo) [inline emlog.h]: Maybe mention additionally that it's for protection from concurrent readers/writers?

File: emlog.c (hunk 2 above)
Comment (megachips) [reply emlog.c]: Line 300. Thanks.

File: emlog.c (hunk 1 above)
Comment (megachips) [reply emlog.c]: Line 300. Thanks.

Comment (ahippo) [reply emlog.c]: No, I mean in case of `return NULL` (new line 273), there is no `read_unlock()`, right?

File: emlog.c (hunk 2 above)
Comment (ahippo) [reply emlog.c]: Same comment as above.

File: emlog.h [hunk 4]
```diff
@@ -39,7 +39,7 @@ struct emlog_info {
     size_t write_point;         /* Offset in circ. buffer of newest data */
     loff_t offset;              /* Byte number of read_point in the stream */
     struct emlog_info *next;
-    rwlock_t rwlock;            /* per e-info rwlock */
+    rwlock_t rwlock;            /* Per e-info rwlock. This is to protect against concurrent readers/writers */
```
Comment (ahippo) [inline emlog.h]: Thank you!

**Comment by megachips** [comment]
Good spotting! Thank you. Have added the necessary unlocks.

**Comment by megachips** [comment]
On a side note, on my local repo I've made a change to emlog module and corresponding mkemlog to allow dynamic sizing of max_emlog_size via module_param - still defaults to 1024. For my application I have max_emlog_size set to 16384 as I was losing data when set at 1Mb. Let me know if you want a pull request. Cheers.

**Comment by ahippo** [comment]
> Good spotting! Thank you. Have added the necessary unlocks.

Thank you so much for your work!
I've just merged it.

> On a side note, on my local repo I've made a change to emlog module and corresponding mkemlog to allow dynamic sizing of max_emlog_size via module_param - still defaults to 1024. For my application I have max_emlog_size set to 16384 as I was losing data when set at 1Mb. Let me know if you want a pull request. Cheers.

Oh, that's a nice feature!
It'd be really cool to merge it!
So, yeah, please, send a pull request in.

**Pull request merged**